[torch export][serialize] create a more compact stacktrace format for serialization

strisunshinewentingwang opened the pull request:
Summary:
- we want fx nodes' stack trace format to be backward compatible and same as before in the program we export
- however in the serialized format, we would want to show a more compact stack_trace format, otherwise the nodes attributes are dominated by stack traces
- the diff implements the minimal in serialization process to dedupe node stack traces by resorting to a fileinfo_list and a filename_to_abbrev map, so we can use index to represent filenames, use lineno to represent lines.

Test Plan:
# llm
base on D54497918
```
buck2 run @//mode/dev-nosan fbcode//executorch/examples/models/llama2:export_llama -- -c ~/stories110M.pt -p ~/params.json
```
set up breakpoint after serialization/deserialization
- serialize
```
(Pdb) v_meta = [n.meta for n in exported_program.graph_module.graph.nodes]
(Pdb) paste_client.create_phabricator_paste_object(paste_creation_client_id=1093956601162697, content=str(v_meta)).number
1193647450
(Pdb) json_program = json.dumps(_dataclass_to_dict(serialized_graph.co_fileinfo_ordered_list),cls=EnumEncoder)
(Pdb) json_bytes = json_program.encode('utf-8')
(Pdb) paste_client.create_phabricator_paste_object(paste_creation_client_id=1093956601162697, content=str(json_bytes)).number
1193604333
(Pdb) sys.getsizeof(json_bytes)
3846
(Pdb) compressed_bytes = zstd.ZstdCompressor().compress(json_bytes)
(Pdb) sys.getsizeof(compressed_bytes)
1139
```
in P1193647450 (before serialization), search for `stack_trace`
in P1193604333 (after serialization), search for `stack_trace` and `co_fileinfo_ordered_list`

[note: didn't do compression in this diff since the size is pretty small and it adds complexity if we do compression]
- deserialize
```
(Pdb) v_meta = [n.meta for n in deserialized_exported_program.graph_module.graph.nodes]
(Pdb) paste_client.create_phabricator_paste_object(paste_creation_client_id=1093956601162697, content=str(v_meta)).number
1193629435
```
in P1193629435, search for `stack_trace`

# ads

Differential Revision: D54654443


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

angelayi commented on torch/_export/serde/serialize.py:
```diff
@@ -1355,7 +1404,12 @@ def deserialize_graph(self, serialized_graph: Graph) -> torch.fx.Graph:
+        self,
+        serialized_node: Node,
+        target: Callable,
+        co_fileinfo_ordered_list: List[CoFileInfo],
```
instead of passing co_fileinfo_ordered_list through the functions, maybe it's better to store it in GraphDeserializer so that you can get it from self.co_fileinfo_ordered_list

strisunshinewentingwang replied:
this is how we construct the GraphModuleDesrializer: 
```
        res = (
            GraphModuleDeserializer()
            .deserialize(
                exported_program.graph_module,
                state_dict,
                constants,
                symbol_name_to_range,
            )
        )
```
do you suggest me to pass in exported_program.graph_module.graph (or co_fileinfo_ordered_list of the graph) as a param to the init of GraphModuleDeserializer?

angelayi replied:
No, like somewhere in [`deserialize_graph`](https://fburl.com/code/hhbc1yac) you can store graph.co_fileinfo_ordered_list into self.co_fileinfo_ordered_list. That way you don't have to pass this variable everywhere else

strisunshinewentingwang replied:
gotcha, yea we can do that, tho I feel this also creates some internal state that is less obvious? I can try it

angelayi reviewed (commented):
cc @zhxchen17 what do you think? Context is in [T173105858](https://www.internalfb.com/tasks?t=T173105858)

zhxchen17 commented on torch/_export/serde/serialize.py:
```diff
@@ -8,8 +9,11 @@
+import textwrap
 import typing
-import copyreg
+import warnings
```
not used?

zhxchen17 commented on torch/_export/serde/schema.py:
```diff
@@ -187,6 +187,26 @@ class Node:
+    file_name: str
+    line_info_map: Dict[str, CoLineInfo] # lineno is str in traceback.FrameSummary
+
+@dataclass
+class Frame:
+    filename: str
+    lineno: int
+    name: str
+    context: str
+
+@dataclass
+class FrameList:
+    items: List[Frame]
```
These data types seem to be helper structures belonging to serialize.py, better to move them there.

zhxchen17 reviewed (changes requested):
thanks for the contribution! overall looks okay to me.

angelayi reviewed (dismissed):
lgtm! thanks!

strisunshinewentingwang committed:
[torch export][serialize] create a more compact stacktrace format for serialization (#121675)

Summary:

- we want fx nodes' stack trace format to be backward compatible and same as before in the program we export 
- however in the serialized format, we would want to show a more compact stack_trace format, otherwise the nodes attributes are dominated by stack traces 
- the diff implements the minimal in serialization process to dedupe node stack traces by resorting to a fileinfo_list and a filename_to_abbrev map, so we can use index to represent filenames, use lineno to represent lines.

Reviewed By: angelayi

Differential Revision: D54654443

strisunshinewentingwang commented:
@pytorchbot merge -f “landed internally"

strisunshinewentingwang commented:
@pytorchbot merge -f "landed internally"

izaitsevfb commented:
@pytorchbot merge -f "landed internally"
